icu4c/source/common/locid.cpp

```diff
@@ -504,34 +508,1104 @@ Locale::operator==( const   Locale& other) const
+                 (uprv_strlen(replacement) == 4 &&
+                  replacement[0] >= '0' &&
+                  replacement[0] <= '9'));
+        if (!same(variant, replacement)) {
+            variants.setElementAt((void*)replacement, i);
+            // Special hack to handle hepburn-heploc => alalc97
+            if (uprv_strcmp(variant, "heploc") == 0) {
+                for (int32_t j = 0; j < variants.size(); j++) {
+                     if (uprv_strcmp((const char*)(variants.elementAt(j)),
+                                     "hepburn") == 0) {
+                         variants.removeElementAt(j);
+                     }
+                }
+            }
+            return true;
+        }
+    }
+    return false;
+}
+
+CharString&
+AliasReplacer::outputToString(
+    CharString& out, UErrorCode status)
+{
+    out.append(language, status);
+    if (notEmpty(script)) {
+        out.append(SEP_CHAR, status)
+            .append(script, status);
+    }
+    if (notEmpty(region)) {
+        out.append(SEP_CHAR, status)
+            .append(region, status);
+    }
+    if (variants.size() > 0) {
+        if (!notEmpty(script) && !notEmpty(region)) {
+          out.append(SEP_CHAR, status);
+        }
+        variants.sort([](UElement e1, UElement e2) -> int8_t {
+            return uprv_strcmp(
+                (const char*)e1.pointer, (const char*)e2.pointer);
+        }, status);
+        for (int32_t i = 0; i < variants.size(); i++) {
+             U_ASSERT(((const char*)(
+                 (UVector*)variants.elementAt(i)))[0] != NULL_CHAR);
+             char uppercaseVariant[256];
+             uprv_strncpy(
+                 uppercaseVariant,
+                 (const char*)((UVector*)variants.elementAt(i)),
+                 sizeof(uppercaseVariant)-1);
+             T_CString_toUpperCase(uppercaseVariant);
+             out.append(SEP_CHAR, status)
+                 .append(uppercaseVariant, status);
+        }
+    }
+    if (notEmpty(extensions)) {
+        CharString tmp("und_", status);
+        tmp.append(extensions, status);
+        Locale tmpLocale(tmp.data());
+        // only support x extension inside CLDR for now.
+        U_ASSERT(extensions[0] == 'x');
+        out.append(tmpLocale.getName() + 1, status);
     }
     return out;
 }
 
+bool
+AliasReplacer::replace(const Locale& locale, CharString& out, UErrorCode status)
+{
+    data = AliasData::singleton(status);
+    if (U_FAILURE(status)) {
+        cleanup();
+        return false;
+    }
+    out.clear();
+    language = locale.getLanguage();
+    if (!notEmpty(language)) {
+        language = nullptr;
+    }
+    script = locale.getScript();
+    if (!notEmpty(script)) {
+        script = nullptr;
+    }
+    region = locale.getCountry();
+    if (!notEmpty(region)) {
+        region = nullptr;
+    }
+    const char* variantsStr = locale.getVariant();
+    const char* extensionsStr = uprv_strchr(locale.getName(), '@');
```

Comment:
This will not work in EBCDIC because different EBCDIC codepages use different byte values for the at sign.
I know there is a helper function in the uloc*.cpp files for this. Please look and call that.

Reply:
ok I use ULOC_KEYWORD_SEPARATOR  but '@' is everywhere. If you really care about fixing for EBCDIC then please file a separate bug for that. I don't think we should snowball the required fix for that into this PR.

Reply:
use locale_getKeywordsStart now